gammapy/cube/ring.py

```diff
@@ -131,122 +110,139 @@ def kernels(self, image):
-        result : dict of `~gammapy.maps.WcsNDMap`
-            Result sky maps.
+        cubes : dict of `~gammapy.maps.WcsNDMap`
+            Dictionary containing `counts_off`, `acceptance` and `acceptance_off` cubes.
         """
-        required = ["counts", "background", "exclusion"]
-        counts, exposure_on, exclusion = [images[_] for _ in required]
-
-        if not counts.geom.is_image:
-            raise ValueError("Only 2D maps are supported")
 
+        counts = dataset.counts
+        background = dataset.background_model.map
         kernels = self.kernels(counts)
-        cubes = {
-            "exposure_on": self._exposure_on_cube(exposure_on, kernels),
-            "exposure_off": self._exposure_off_cube(exposure_on, exclusion, kernels),
-            "off": self._off_cube(counts, exclusion, kernels),
-        }
-        cubes["alpha_approx"] = self._alpha_approx_cube(cubes)
 
-        exposure_off, off = self._reduce_cubes(cubes)
-        alpha = exposure_on.data / exposure_off
+        cubes = {}
+        cubes["counts_off"] = scale_cube(
+            (counts.data * exclusion.data)[0, Ellipsis], kernels
+        )
+        cubes["acceptance_off"] = scale_cube(
+            (background.data * exclusion.data)[0, Ellipsis], kernels
+        )
+
+        scale = background.geom.pixel_scales[0].to("deg")
+        theta = self.parameters["theta"] * scale
+        tophat = Tophat2DKernel(theta.value)
+        tophat.normalize("peak")
+        acceptance = background.convolve(tophat.array)
+        acceptance_data = acceptance.data[0, Ellipsis]
+        cubes["acceptance"] = np.repeat(
+            acceptance_data[Ellipsis, np.newaxis], len(kernels), axis=2
+        )
+
+        return cubes
 
-        # set data outside fov to zero
-        not_has_exposure = ~(exposure_on.data > 0)
-        for data in [alpha, off, exposure_off]:
-            data[not_has_exposure] = 0
+    def run(self, dataset, exclusion):
+        """Run adaptive ring background maker
 
-        background = alpha * off
+        Parameters
+        ----------
+        dataset : `~gammapy.cube.fit.MapDataset`
+            Input map dataset.
+        exclusion : `~gammapy.maps.WcsNDMap`
+            Exclusion mask for regions with known gamma-ray emission.
 
-        return {
-            "exposure_off": counts.copy(data=exposure_off),
-            "off": counts.copy(data=off),
-            "alpha": counts.copy(data=alpha),
-            "background_ring": counts.copy(data=background),
-        }
+        Returns
+        -------
+        dataset_on_off : `~gammapy.cube.fit.MapDatasetOnOff`
+            On off dataset.
+        """
+        cubes = self.make_cubes(dataset, exclusion)
+        acceptance, acceptance_off, counts_off = self._reduce_cubes(cubes, dataset)
 
+        not_has_off_acceptance = acceptance_off.data <= 0
+        acceptance_off.data[not_has_off_acceptance] = np.nan
 
-class RingBackgroundEstimator:
+        fft_noise_threshold = 1e-6
```

Comment:
This hard-coded threshold seem problematic to me. I guess in general the `acceptance_off` is zero, where the ring completely falls into an exclusion region, no? What if we ring-convolve the `exclusion_mask` as well and derive the off mask from this? I guess we can also use `scripy.ndimage.morphology`. Here is a minimal example:
```python3
from astropy.convolution import Ring2DKernel
from scipy.ndimage.morphology import binary_erosion, binary_dilation
from gammapy.maps import Map
from regions import CircleSkyRegion
from astropy import units as u
from astropy.coordinates import SkyCoord

binsz= 0.02
center = SkyCoord("0 deg", "0 deg")
circle = CircleSkyRegion(center=center, radius=0.4 *u.deg)

exclusion_mask = Map.create(width=2, binsz=binsz)
exclusion_mask.data = exclusion_mask.geom.region_mask([circle], inside=True)

ring = Ring2DKernel(10, width=5)
ring.normalize("peak")

data = binary_erosion(exclusion_mask.data, structure=ring.array)
mask = exclusion_mask.copy(data=data)
```